snuba/datasets/all_events.py

```diff
@@ -0,0 +1,165 @@
+
+# Based on query.get_all_referenced_columns()
+def get_condition_columns(query: Query) -> Sequence[Any]:
+    col_exprs: MutableSequence[Any] = []
+    # Conditions need flattening as they can be nested as AND/OR
+    if query.get_conditions():
+        flat_conditions = list(
+            chain(*[[c] if is_condition(c) else c for c in query.get_conditions()]))
+        col_exprs.extend([c[0] for c in flat_conditions])
+
+    # Return the set of all columns referenced in any expression
+    return set(chain(*[columns_in_expr(ex) for ex in col_exprs]))
+
+
+def detect_dataset(query: Query) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
+
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    if [col for col in TRANSACTIONS_ONLY_COLUMNS if col in get_condition_columns(query)]:
+        return TRANSACTIONS
+
+    # Use events by default
+    return EVENTS
+
+
+class AllEventsProcessor(QueryProcessor):
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        """
+        Switches the data source from the default (Events) to the transactions
+        table if a transaction specific column is detected.
+        Sets all the other columns to none
+
+        """
+        detected_dataset = detect_dataset(query)
+
+        source = get_dataset(detected_dataset) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
+        query.set_data_source(source)
+
+        if detected_dataset == TRANSACTIONS:
+            with_columns = [
+                ("'transaction'", 'type'),
+                ('finish_ts', 'timestamp'),
+                ('user_name', 'username'),
+                ('user_email', 'email'),
+            ] + [('NULL', col) for col in EVENTS_ONLY_COLUMNS]
+
+        else:
+            with_columns = [('NULL', col) for col in TRANSACTIONS_ONLY_COLUMNS]
+
+        query.set_with(with_columns)
+
+
+class AllEventsSchema(Schema, ABC):
+    def get_data_source(self) -> RelationalSource:
+        """
+        This is a placeholder, we switch out the data source in the processor
+        depending on the detected dataset
+        """
+        return get_dataset(EVENTS) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
```

Comment:
Not sure I follow this:
this is supposed to be the abstract schema for the virtual table sentry will see. Shouldn't it have all columns coming from both transactions and events together (with the nullability tweaked accordingly) ?

Reply:
I changed this to just return None. In the current implementation I wanted to avoid introducing a new data source since this dataset only switches between the existing events and transactions sources right now. These lines here are just placeholder, since the dataset actually gets picked in the processor depending on what dataset is detected.

Are you proposing doing the dataset detection inside DataSource instead of the query processor and then making DataSource pick what table gets selected?

Reply:
We still need an abstract schema (see the get_columns method).
While it is true that today we would not be using it, soon we will use it to parse the column queries and to validate the query itself.
Also it would document how this dataset works and what it returns, otherwise people will have to guess by reading the processors.

>Are you proposing doing the dataset detection inside DataSource instead of the query processor and then making DataSource pick what table gets selected?
No, please let's keep this in the processor. Or at least, if you want to put the decision logic in the DataSource I am fine as long as it is triggered by a query processor so we respect the rule based abstraction. But in that case such change would not be very useful.

Reply:
No I definitely would prefer to keep it in the processor. I was just wondering why you were proposing to provide a fully detailed DataSource that is not used. If it's just for documentation now that's fine.